frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -832,6 +858,110 @@ pub mod pallet {
+            let last_allowed_era = current_era.saturating_sub(Self::history_depth());
+            let start_from_era = last_claim_era.max(last_allowed_era);
+            if (start_from_era > last_claim_era) {
+                // TODO collect all unclaimed rewards and send to Treasury pallet
+            }
+
+            // for the first claimable era "start_from_era", this storage item must be in place!
+            let mut contract_stake_prev = Self::contract_era_stake(&contract_id, &start_from_era)
+                .ok_or(Error::<T>::UnknownStartStakingData)?;
+
+            // for any era after start_from_era, the ContractEraStake is present only if there
+            // was a change in staking amount. If it is not present we process last recorded ContractEraStake
+            for era in start_from_era..current_era {
+                let total_era = Self::get_era_total(era).ok_or(Error::<T>::UnknownEraReward)?;
+                let contract_stake =
+                    Self::contract_era_stake(&contract_id, era).unwrap_or(contract_stake_prev);
+
+                // this contract's part in whole era reward
+                let contract_part = Perbill::from_rational(
+                    Self::balance_to_u64(total_era.rewards).unwrap_or(1),
+                    Self::balance_to_u64(total_era.staked).unwrap_or(1),
+                );
+
+                let contract_era_reward =
+                    contract_part * Self::balance_to_u64(contract_stake.total).unwrap_or(0);
+
+                // divide reward between stakers and the developer of the contract
+                let contract_staker_reward =
+                    Perbill::from_rational((100 - T::DeveloperRewardPercentage::get()) as u64, 100)
+                        * contract_era_reward;
+                let contract_developer_reward =
+                    Perbill::from_rational(T::DeveloperRewardPercentage::get() as u64, 100)
+                        * contract_era_reward;
+
+                // make payout to the stakers and the developer
+                Self::payout_stakers2(&contract_stake, contract_staker_reward);
+                Self::payout_developer(developer.clone(), contract_developer_reward);
+
+                // store current record in case next era has no record of changed stake amount
+                contract_stake_prev = contract_stake;
+            }
+
+            // Remove all previous records of staking for this contract,
+            // they have already been processed and won't be needed anymore.
+            ContractEraStake::<T>::remove_prefix(&contract_id, None);
+
+            // move contract pointers to current era
+            ContractLastClaimed::<T>::insert(&contract_id, current_era);
+            ContractLastStaked::<T>::insert(&contract_id, current_era);
+
+            // create contract era stake data in current era for further staking and claiming
+            ContractEraStake::<T>::insert(&contract_id, current_era, contract_stake_prev);
```

Comment:
Please move this directly below the line where you delete all content related to this contract_id so it's clear that only the last era info is kept.